### ISSUE TITLE
TPM Support Status

### ISSUE THREAD

**Issue by erenoglu** [arnoldthebat/chromiumos#274]
Dear Arnoldthebat, just wondering if there were any developments on getting TPM support working, either by an emulated tpm or enabling the use of the hardware one?

As we discussed earlier, I can't take ownership of the tpm2 (TIS) on my lenovo x1 carbon 5th gen.

**Comment by Crescendo-BLYAT** [comment]
@arnoldthebat also, is there any news for Special v74 build?

Don't forget to include swtpm.tar for your builds to fix this TPM issue as mentioned sometimes ago.

**Comment by erenoglu** [comment]
Thanks @Crescendo-BLYAT, My understanding from the swtpm.tar is that it fixes the login issue. I don't have a login issue on my X1 Carbon 5th gen (tis tpm2.0). 

Can you confirm that you can import & bind certificates with swtpm.tar?  You can try to import any Certificate Authority.

**Comment by Crescendo-BLYAT** [comment]
@erenoglu yes & yes.... also fixed the infamous empty flags page & device ownership....

**Comment by erenoglu** [comment]
Yes, yesterday I tried as well. But I think I needed to blacklist tpm_tis module as well as my kernel insisted on loading it even if it was disabled in bios

**Comment by arnoldthebat** [comment]
Hi

Since TPM is working in later hardware, its enabled by default in all
builds. Bear in mind, you ned to clear TPM in BIOS and then let ChromiumOS
claim it on boot. Then device ownership and flags etc work (on my kit
anyway)

Software TPM is too disruptive to deploy to the main special build now I
have investigated it, so Ill will be looking to create a variant to allow
for software TPM only.

**Comment by arnoldthebat** [comment]
Hi

Ill be releasing R74 imminently. Software TPM will need to be a variant
build however since I dont want to disable hardware TPM completely.

**Comment by erenoglu** [comment]
thanks @arnoldthebat , I'm really wondering what's wrong with my HW TPM since it's impossible to get ChromeOS or ChromiumOS or FydeOS own it (after clearing it in BIOS or Windows), although Windows & Linux can own it just fine :( 
swtpm worked on my system yesterday so that looks like the only way out for the moment

**Comment by arnoldthebat** [comment]
Bear with me a little longer then in that case, and Ill get a variant build with SWTPM only...

**Comment by Crescendo-BLYAT** [comment]
> Bear with me a little longer then in that case, and Ill get a variant build with SWTPM only...

couldn't wait for this build as my laptop's hwTPM is crazy...
it got sth to do with Acer's implementation of TPM & secure boot...

Chromium unable to acquire it even tho I already cleared the TPM's data via BIOS....

Thank you so much...

**Comment by arnoldthebat** [comment]
All. Try this test build please: https://chromium.arnoldthebat.co.uk/.archive//chromiumos_image.bin.gz

SWTPM appears to be loading and the service _vtpm shows as running. Its likely I have missed something since certs and chrome://flags are still not loading on my test laptop, so if you could have a review, I would appreciate it. Otherwise Ill try to do some debug this weekend if I get time

**Comment by Crescendo-BLYAT** [comment]
nice, I'll help you test this one.... :)
Thank you so much.

**Comment by erenoglu** [comment]
Hi @arnoldthebat , check if you have a hardware tpm and if it's module is loaded. If it's loaded, it captures /dev/tpm0 device and the swtpm gets /dev/tpm1

I needed to 1) disable HW tpm from BIOS 2) disable module loading before I could get swtpm to work.
Also you need to have the links in /usr/lib64 for libswtp.... and libtpm... (as in the chromfy script):

also you need to start from fresh state partition, at first boot, in 2nd or 3rd screen where google asks if it can send some diagnostic data, there's a link above to show the password. If that link shows you a password, it's working. 

Attached are my related files. (had to add .txt extension to them to upload, you shall remove that)
[vtpm.conf.txt](https://github.com/arnoldthebat/chromiumos/files/3240846/vtpm.conf.txt)
[tis.conf.txt](https://github.com/arnoldthebat/chromiumos/files/3240848/tis.conf.txt)

**Comment by eladavron** [comment]
Hi, so it doesn't seem to work in Camd64OS_R76-12239.B-Special on a Surface GO 128GB (which has the TPM2.0 module).
Chrome Flags doesn't work, and I *think* YouTube Android app doesn't work because of it too (though I'm not sure it's this reason).
Anything I can do to help with solving this?
